utils_cv/detection/dataset.py

```diff
@@ -0,0 +1,220 @@
+import xml.etree.ElementTree as ET
+from PIL import Image
+
+from .plot import display_bounding_boxes, plot_grid
+from .model import get_transform
+
+
+class DetectionDataset(object):
+    """ An object detection dataset.
+    """
+
+    def __init__(
+        self,
+        root: Union[str, Path],
+        transforms: object = None,
+        image_folder: str = "images",
+        annotation_folder: str = "annotations",
+    ):
+        """ initialize dataset
+
+        This class assumes that the data is formatted in two folders:
+            - annotation folder which contains the Pascal VOC formatted
+              annotations
+            - image folder which contains the images
+
+        Args:
+            root: the root path of the dataset containing the image and
+            annotation folders
+            transforms: the transformations to apply
+            image_folder: the name of the image folder
+            annotation_folder: the name of the annotation folder
+        """
+
+        self.root = Path(root)
+        self.transforms = transforms
+        self.image_folder = image_folder
+        self.annotation_folder = annotation_folder
+
+        self.ims = list(sorted(os.listdir(self.root / self.image_folder)))
+        self.annotations = list(
+            sorted(os.listdir(self.root / self.annotation_folder))
+        )
+        self.categories = self._get_categories()
+
+    def _get_categories(self) -> List[str]:
+        """ Parses all Pascal VOC formatted annoatation files to extract all
+        possible categories. """
+        categories = ["__background__"]
+        for annotation_path in self.annotations:
+            annotation_path = self.root / "annotations" / str(annotation_path)
+            tree = ET.parse(annotation_path)
+            root = tree.getroot()
+            objs = root.findall("object")
+            for obj in objs:
+                category = obj[0]
+                assert category.tag == "name"
+                categories.append(category.text)
+        return list(set(categories))
+
+    def get_image_features(
+        self, idx: int = None, rand: bool = False
+    ) -> Tuple[List[List[int]], List[str], str]:
+        """ Choose and get image from dataset.
+
+        This function returns all the data that is needed to effectively
+        visualize an image from the dataset.
+
+        Args:
+            idx: The image index to get the features of
+            rand: randomly select image
+
+        Raises:
+            Exception is idx is not None and rand is set to True
+
+        Returns:
+            A tuple of boxes, categories, image path
+        """
+        if (idx is not None) and (rand is True):
+            raise Exception("idx cannot be set if rand is set to True.")
+        if idx is None and rand is False:
+            rand = True
+        if rand:
+            idx = randrange(len(self.ims))
+        boxes, labels, im_path = self._get_im_data(idx)
```

Comment:
Is label here an integer, and category the actual class name? Is that always the case? Nice point for a BBox class, there we could have a "class_name" and "class_id" fields

Reply:
yes, i've been using label to indicate int and category to indicate classname, bbox class wuodl help for sure.

Reply:
Vanja pointed out to me sometime ago that we use label/class inconsistenly in classification. Her point is that "classes" should refer to the set of distinct concepts ie classes = set("dog", "cat", "person") while label is an instation, e.g. for a given image where 5 objects were found the returned labels could be labels=["dog", "dog", "cat", "person", "cat"].
So given that, should we use e.g. "labels" if its the names (which is the more common case I am guess), and "label_ids" if its their index?